ethrpc/ens.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 Weald Technology Trading
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ethrpc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/0xsequence/ethkit/go-ethereum/common"
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/net/idna"
+)
+
+// TODO: Add a cachestore to cache the results of the ENS lookups
+
+const ENSContractAddress = "0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e"
+
+var p = idna.New(idna.MapForLookup(), idna.StrictDomainName(false), idna.Transitional(false))
+
+func ResolveEnsAddress(ctx context.Context, ens string, provider *Provider) (common.Address, bool, error) {
+	// check if it's an address
+	ensAddress := common.HexToAddress(ens)
+	if ensAddress.Hex() == ens {
+		return ensAddress, true, nil
+	}
+
+	chainId, err := provider.ChainID(ctx)
+	if err != nil {
+		return common.Address{}, false, fmt.Errorf("ethrpc: failed to get chainId of the passed provider")
+	}
+	if chainId.Int64() != 1 {
```

Comment:
lets remove this assumption, and remove "chainId" parameter altogether. The idea here is, ENS resolver takes a name/address and resolves to the address. It will always check the mainnet contract. It's irrelevant which chain is being queried, as I believe ENS is only deployed on mainnet, so that is the source of truth we will query. Addresses can be the same across chains, like in the case of Sequence

Reply:
We are enforcing users to pass a mainnet provider with their node, which we will use to query ENS. Shouldn't we check the passed provider is mainnet node?

Reply:
Nice point. Can you make it accept ethproviders.Providers instead, then grab the provider from mainnet, if it’s not available, then error to say mainnet provider is unavailable

Reply:
Using `ethprovider.Providers` here will cause import cycle as we use `ethrpc` in `ethproviders` package. Since we have to send only one provider, what's the use of `ethprovider.Providers` ?